hs-bindgen/src/HsBindgen/C/Tc/Macro.hs

```diff
@@ -0,0 +1,1121 @@
+      let !tvs' = goTy bound tvs argTy
+      in case NE.nonEmpty mbArgTys of
+        Nothing     -> goTy bound tvs resTy
+        Just argTys -> goFunTy bound tvs' argTys resTy
+
+tyVarSet :: [ TyVar ] -> IntSet
+tyVarSet = IntSet.fromList . map ( uniqueInt . tyVarUnique )
+
+newtype Subst tv = Subst { substMap :: IntMap ( tv, Type Ty ) }
+  deriving newtype Monoid
+instance Functor Subst where
+  fmap f ( Subst s ) = Subst $ fmap ( first f ) s
+instance Show k => Semigroup ( Subst k ) where
+  ( Subst s1 ) <> sub2@( Subst s2 ) =
+    Subst $ IntMap.unionWithKey ( substClashErr "Semigroup Subst" )
+              s2
+              ( IntMap.map ( \ ( nm, ty ) -> ( nm, applySubst sub2 ty ) ) s1 )
+instance Show k => Show ( Subst k ) where
+  show ( Subst s ) = "{ " ++ intercalate ", " ( map f $ IntMap.elems s ) ++ " }"
+    where
+      f ( tv, ty ) = show tv ++ " |-> " ++ show ty
+
+
+addOneToSubst :: HasCallStack => TyVar -> Type Ty -> Subst TyVar -> Subst TyVar
+addOneToSubst tv ty s = s <> mkSubst [ ( tv, ty ) ]
+
+mkSubst :: HasCallStack => [ ( TyVar, Type Ty ) ] -> Subst TyVar
+mkSubst = Subst
+        . IntMap.fromListWithKey ( substClashErr "mkSubst " )
+        . map ( \ ( tv, ty ) -> ( uniqueInt ( tyVarUnique tv ), ( tv, ty ) ) )
+
+
+substClashErr :: ( Show a, HasCallStack ) => String -> Int -> a -> a -> a
+substClashErr str i ty1 ty2 =
+  error $
+    unlines
+      [ str ++ ": incoherent substitution"
+      , "TyVar with unique " ++ show ( Unique i ) ++ " mapped to two different types"
+      , "ty1: " ++ show ty1
+      , "ty2: " ++ show ty2
+      ]
+
+lookupSubst :: TyVar -> Subst tv -> Maybe ( Type Ty )
+lookupSubst tv ( Subst s ) = fmap snd $ IntMap.lookup ( uniqueInt $ tyVarUnique tv ) s
+
+applySubst :: forall ki tv. HasCallStack => Subst tv -> Type ki -> Type ki
+applySubst = goTy
+  where
+    goTy :: forall ki'. Subst tv -> Type ki' -> Type ki'
+    goTy subst = \case
+      ty@( TyVarTy tv ) ->
+        case IntMap.lookup ( uniqueInt $ tyVarUnique tv ) ( substMap subst ) of
+          Nothing  -> ty
+          Just ( _, ty' ) -> ty'
+      FunTy args res ->
+        FunTy ( fmap ( applySubst subst ) args ) ( applySubst subst res )
+      TyConAppTy tc tys ->
+        TyConAppTy tc $ fmap ( applySubst subst ) tys
+
+{-------------------------------------------------------------------------------
+  Constraints & errors
+-------------------------------------------------------------------------------}
+
+type FunName = Either CName MFun
+showFunName :: FunName -> String
+showFunName = \case
+  Left ( CName f ) -> Text.unpack f
+  Right f -> show f
+
+-- | Why did we emit a constraint?
+data CtOrigin
+  = AppOrigin !FunName
+  | InstOrigin !FunName
+  | IntLitOrigin !( Literal Integer )
+  | FloatLitOrigin !Double
+  deriving stock Generic
+  deriving anyclass PrettyVal
+
+instance Show CtOrigin where
+  show = \case
+    AppOrigin fun ->
+      "In an application of '" ++ showFunName fun ++ "'."
```

Comment:
I'm not a fan of non-derived like `Show` instances. ping @edsko

Reply:
Fair point. I made these into separate functions and derived the Show instance.